### ISSUE TITLE
Unable to log in with 2FA

### ISSUE THREAD

**Issue by Phyks** [stom79/mastalab#357]
Hi,

I'm using Mastalab on an (old) Android phone stuck with 4.2.2. I'm using 2FA with my Mastodon account and clicked the link for 2FA authentication. My problem is that apparently this link has to open in the default Android browser and I get an error stating that no connection could be made.

When I'm opening the same URL with Firefox for Android, everything works fine, but it does not send back the result to Mastalab so I cannot set up my account :/

My guess for now (quite a wild guess actually) is that my instance has some restrictive setup for its HTTPS version (in terms of ciphers) and that these are not supported by the built-in Android browser. Firefox on the contrary is still up to date and works fine with this website. Do you think it could be possible to let this link be opened in another browser? Otherwise, would it be possible to manually set up my account by providing an OAuth token directly?
---

Instance: mastodon.tetaneutral.net

Mastalab version: latest from f-droid (1.8.3)

Android version: 4.2.2

Thanks for all the work on this wonderful Mastodon client!

**Comment by stom79** [comment]
Yes, I didn't patch the security provider for the built-in browser. But, you can click on the three dots and select "Custom tabs" that will allow you to pick up an external browser. Let me know if it's ok.

**Comment by Phyks** [comment]
Indeed, got it working. Actually, I did not notice the link was longer than the actual error view, so when I tried to copy it in Firefox, I missed part of it. Got it working now, thanks!

Not sure but maybe clicking "2FA" could trigger a prompt "which app to go on with?" with any browser available on the system?

**Comment by stom79** [comment]
Why did you copy a link in FF? Custom tabs initially ask you what external browser you want to use (so every browsers in your device should be suggested - I guess it's what you asked before). Then, the link is automatically opened with this browser and once authenticated, the app should take back the process.

I tested with FF and I got an issue with multiple tabs that where not using the right client id. I fixed it.

**Comment by Phyks** [comment]
Hmmm, not sure what the custom tabs are. When I start Mastalab (first start), I click the "Two-factor auth" link at the bottom of the onboarding connection screen. Then, this opens a webview with what should be the connection prompt on my Mastodon instance, to get the OAuth token. Problem is that this webview cannot work on my Android version (I guess due to it being too old and the webview not being compatible with the HTTPS from my instance).

In this browser page, I cannot do any actions beside selecting text. Then, I selected and copied the URL appearing in the error message and put it in Firefox.

**Comment by stom79** [comment]
OK. I didn't understand how you got that link :)
And yes, the toast message is truncated, so you can't use it. I really need
to improve the UX by removing these messages.
For custom tabs, there is this link:

https://developer.chrome.com/multidevice/android/customtabs.

It works without problems with firefox

Hmmm, not sure what the custom tabs are. When I start Mastalab (first
start), I click the "Two-factor auth" link at the bottom of the onboarding
connection screen. Then, this opens a webview with what should be the
connection prompt on my Mastodon instance, to get the OAuth token. Problem
is that this webview cannot work on my Android version (I guess due to it
being too old and the webview not being compatible with the HTTPS from my
instance).

In this browser page, I cannot do any actions beside selecting text. Then,
I selected and copied the URL appearing in the error message and put it in
Firefox.

—
You are receiving this because you were assigned.

Reply to this email directly, view it on GitHub
<https://github.com/stom79/mastalab/issues/357#issuecomment-374030613>, or mute
the thread

**Comment by Phyks** [comment]
> For custom tabs, there is this link: 

Sure, I see this when I click a link in a toot, with a button in the top right corner to expand in a new tab in the browser. But this expand button was not here when setting up 2FA. Not sure why.

And it seems, although expanding to a tab indeed opens it in Firefox, that it is not using Firefox for the render inside the custom tab as is shown by Github not being able to display.

![screenshot_2018-03-19-11-29-46](https://user-images.githubusercontent.com/3856586/37590677-56b74416-2b69-11e8-8c64-9fe34f54e486.png)


Anyways, there seems to be quite a few bugs with customtabs in Android 4.2.2. Not sure if it's really a Mastalab issue in the end. Typically, if I open a link from a toot, press back (I'm now back at the toot view) and press the back button again, I'm back in the customtab instead of being sent back to the timeline. I remember having similar issues with Tusky, so that's why I'm thinking about an Android 4.2.2 bug.

**Comment by stom79** [comment]
The button at the top right allows to open the link with an external
browser. It is not displayed in the login activity.
You should go in settings and disable the built-in in browser, then uncheck
custom tabs. It should automatically open links with Firefox after choosing
it the first time.

Le 19 mars 2018 11:36 AM, "Lucas Verney" <notifications@github.com> a
écrit :

For custom tabs, there is this link:

Sure, I see this when I click a link in a toot, with a button in the top
right corner to expand in a new tab in the browser. But this expand button
was not here when setting up 2FA. Not sure why.

And it seems, although expanding to a tab indeed opens it in Firefox, that
it is not using Firefox for the render inside the custom tab as is shown by
Github not being able to display.

[image: screenshot_2018-03-19-11-29-46]
<https://user-images.githubusercontent.com/3856586/37590677-56b74416-2b69-11e8-8c64-9fe34f54e486.png>

Anyways, there seems to be quite a few bugs with customtabs in Android
4.2.2. Not sure if it's really a Mastalab issue in the end. Typically, if I
open a link from a toot, press back (I'm now back at the toot view) and
press the back button again, I'm back in the customtab instead of being
sent back to the timeline. I remember having similar issues with Tusky, so
that's why I'm thinking about an Android 4.2.2 bug.

—
You are receiving this because you were assigned.

Reply to this email directly, view it on GitHub
<https://github.com/stom79/mastalab/issues/357#issuecomment-374168371>, or mute
the thread

**Comment by Phyks** [comment]
> You should go in settings and disable the built-in in browser, then uncheck custom tabs. It should automatically open links with Firefox after choosing it the first time.

Firefox is already the default browser on my device. When going to settings, I cannot disable the built-in browser (already cleared its data and cleared default actions though). Not sure these settings are actually available under AOSP 4.2.2.
